app/controllers/application_controller.rb

```diff
@@ -5,6 +5,136 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "abcdefghijklmnop"
+  end
+
+  helpers do
+    def logged_in?
+      !!session[:user_id]
+    end
+
+    def current_user
+      User.find_by(id: session[:user_id])
+    end
+  end
+
+
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if logged_in?
+      redirect "/tweets"
+    else
+      erb :"users/create_user"
+    end
+  end
+
+  post '/signup' do
+    redirect "/users/signup" if params[:username] == "" || params[:email] == "" || params[:password] == ""
+
+    user = User.new
+    user.username = params[:username]
+    user.email = params[:email]
+    user.password = params[:password]
+    user.save
+    session[:user_id] = user.id
+
+    redirect "/tweets"
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect "/tweets"
+    else
+      erb :"users/login"
+    end
+  end 
+
+  post '/login' do
+    redirect "/users/login" if params[:username] == "" || params[:password] == ""
+    user = User.find_by(username: params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect "/tweets"
+    else
+      redirect "/login"
+    end
+  end
+
+  get '/logout' do
+    session.clear if logged_in?
+    redirect "/login"
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :"users/show"
+  end
+
+  # Tweet related
+
+  get '/tweets' do
+    redirect "/login" if !logged_in?
+
+    @tweets = Tweet.all
+    erb :"tweets/tweets"
+  end
+
+  get '/tweets/new' do
+    redirect "/login" if !logged_in?
+
+    erb :"tweets/new"
+  end
+
+  get '/tweets/:id' do
+    redirect "/login" if !logged_in?
+
+    @tweet = Tweet.find_by(id: params[:id])
+
+    erb :"tweets/show"
+  end
+
+  get '/tweets/:id/edit' do
+    redirect "/login" if !logged_in?
+
+    @tweet = Tweet.find_by(id: params[:id])
+  
+    if @tweet.user.id == current_user.id
+      erb :"tweets/edit" 
+    else
+      redirect :"/tweets"
+    end
+  end
+
+  post '/tweets' do
+    redirect "/tweets/new" if params[:content] == ""
+
+    @tweet = Tweet.create(params)
+    @tweet.user = current_user
+    @tweet.save
+
+    redirect "/tweets/#{@tweet.id}"
+  end
+
+  patch '/tweets/:id' do
+    redirect "/tweets/#{params[:id]}/edit" if params[:content] == ""
+
+    @tweet = Tweet.find_by(id: params[:id])
+    @tweet.update(content: params[:content])
+
+    redirect "/tweets/#{@tweet.id}"
+  end
+
+  delete '/tweets/:id/delete' do
+    redirect "/login" if !logged_in?
+
+    @tweet = Tweet.find_by(id: params[:id])
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.